src/order/spam-protection/google-recaptcha.ts

```diff
@@ -1,23 +1,90 @@
+import { NotInitializedError, NotInitializedErrorType } from '../../common/error/errors';
+
 import GoogleRecaptchaScriptLoader from './google-recaptcha-script-loader';
 
+export interface RecaptchaResult {
+    error?: Error;
+    token?: string;
+}
+
 export default class GoogleRecaptcha {
+    private _event$?: Subject<RecaptchaResult>;
+    private _recaptcha?: ReCaptchaV2.ReCaptcha;
+
     constructor(
-        private googleRecaptchaScriptLoader: GoogleRecaptchaScriptLoader
+        private googleRecaptchaScriptLoader: GoogleRecaptchaScriptLoader,
+        private mutationObserverFactory: MutationObserverFactory
     ) {}
 
-    render(containerId: string, sitekey: string, callbacks: GoogleRecaptchaCallbacks): Promise<void> {
+    load(containerId: string, sitekey: string): Promise<void> {
+        const event$ = new Subject<RecaptchaResult>();
+        this._event$ = event$;
+
         return this.googleRecaptchaScriptLoader.load()
             .then(recaptcha => {
                 recaptcha.render(containerId, {
                     sitekey,
-                    ...callbacks,
+                    size: 'invisible',
+                    callback: () => {
+                        event$.next({
+                            token: recaptcha.getResponse(),
+                        });
+                    },
+                    'error-callback': () => {
+                        event$.next({
+                            error: new Error('Unable to complete spam verification because Recaptcha encountered an error.'),
+                        });
+                    },
                 });
+
+                this._recaptcha = recaptcha;
             });
     }
-}
 
-interface GoogleRecaptchaCallbacks {
-    callback(token: string): void;
-    'error-callback'?(): void;
-    'expired-callback'?(): void;
+    execute(): Observable<RecaptchaResult> {
+        if (!this._event$ || !this._recaptcha) {
+            throw new NotInitializedError(NotInitializedErrorType.SpamProtectionNotInitialized);
+        }
+
+        this._watchRecaptchaChallengeWindow();
+
+        this._recaptcha.execute();
+
+        return this._event$;
+    }
+
+    private _watchRecaptchaChallengeWindow() {
+        const iframeElement = document.querySelector('iframe[title="recaptcha challenge"]');
+
+        if (!iframeElement) {
+            throw new Error('Recaptcha challenge iframe not found.');
+        }
+
+        const iframeContainer = iframeElement.parentElement;
+
+        if (!iframeContainer) {
+            throw new Error('Recaptcha challenge iframe container not found.');
+        }
+
+        const container = iframeContainer.parentElement;
+
+        if (!container) {
+            throw new Error('Recaptcha challenge container not found.');
+        }
+
+        this.mutationObserverFactory.create(() => {
+            if (container.style.visibility === 'hidden') {
+                if (!this._event$) {
```

Comment:
this condition can never happen, so I might not bother throwing an error. Maybe pass the event as an argument to avoid doing this check in here?

Reply:
Sure, i passed the event as an argument.